### PR TITLE
KAFKA-7638: Add support for multiple task creation in one request

### PR DESCRIPTION
**Pull request by stanislavkozlovski**
[JIRA](https://issues.apache.org/jira/browse/KAFKA-7638)
Trogdor Coordinator now supports creation of multiple tasks in one request call.
This is exposed under `/coordinator/tasks/mass_create`
Sample body: 
```json
{"tasks": [{...}, {...}]}
```

Additional changes:
The /tasks/create endpoint (and now mass_create) return 400 with an error message whenever an error is encountered during the task creation. Previously, Trogdor would create the task as "DONE" and return a successful status code, making users manually inspect the tasks' status.
This, unfortunately, has the nasty downside of returning 400 when an unexpected internal server error occurs. 
I think that we can live with this for now and open up another KIP that could address error handling in the Trogdor Coordinator REST API.

### CODE REVIEW

File: tools/src/main/java/org/apache/kafka/trogdor/coordinator/CoordinatorRestResource.java [hunk 1]
```diff
@@ -76,9 +80,16 @@ public CoordinatorStatusResponse status() throws Throwable {
 
     @POST
     @Path("/task/create")
-    public Empty createTask(CreateTaskRequest request) throws Throwable {
+    public Response createTask(CreateTaskRequest request) throws Throwable {
         coordinator().createTask(request);
-        return Empty.INSTANCE;
+        return Response.status(201).entity(Empty.INSTANCE).build();
```
Comment (cmccabe) [inline tools/src/main/java/org/apache/kafka/trogdor/coordinator/CoordinatorRestResource.java:86]: Cool.  A 201 CREATED response makes sense here.

File: tools/src/main/java/org/apache/kafka/trogdor/coordinator/TaskManager.java [hunk 2]
```diff
@@ -294,25 +298,82 @@ private JsonNode getCombinedStatus() {
...
+    }
+
+    /**
+     * Atomically creates and schedules the given tasks. If any task fails creation, none get scheduled.
+     */
+    public void createAndScheduleTasksAtomic(List<TaskDetail> tasks) throws BadRequestException, RequestConflictException {
+        List<ManagedTask> managedTasks = new ArrayList<>();
+        if (tasks.stream().map(task -> task.id).collect(Collectors.toSet()).size() < tasks.size())
+            throw new RequestConflictException("Duplicate task IDs given");
+
```
Comment (cmccabe) [inline tools/src/main/java/org/apache/kafka/trogdor/coordinator/TaskManager.java]: We should also check to make sure there are no invalid empty task IDs.  In that case we should throw an exception and not try to create anything, similar to the above exception...

File: tools/src/main/java/org/apache/kafka/trogdor/coordinator/TaskManager.java [hunk 3]
```diff
@@ -322,42 +383,13 @@ public void createTask(final String id, TaskSpec spec)
...
-            String failure = null;
+        public ManagedTask call() throws Exception {
             try {
-                controller = spec.newController(id);
-            } catch (Throwable t) {
-                failure = "Failed to create TaskController: " + t.getMessage();
+                TaskController controller = spec.newController(id);
+                return new ManagedTask(id, spec, controller, ManagedTaskState.PENDING);
+            } catch (Exception t) {
+                throw new BadRequestException("Failed to create TaskController: " + t.getMessage());
```
Comment (cmccabe) [inline tools/src/main/java/org/apache/kafka/trogdor/coordinator/TaskManager.java]: This exception gets logged and then ignored by `createAndScheduleTasksAtomic`, right?  I think it would be better to keep the original behavior where we create a task, and mark it as done and failed.  Then, at least, the user could see the error message if something failed.  The error would appear in the failed task status.  The user shouldn't have to read the logfile to see the failure message.

File: tools/src/main/java/org/apache/kafka/trogdor/rest/CreateMultipleTasksRequest.java [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+
+/**
+ * A request to the Trogdor coordinator to create mulitple tasks.
+ */
+public class CreateMultipleTasksRequest extends Message {
+    private final List<CreateTaskRequest> tasks;
+
+    @JsonCreator
+    public CreateMultipleTasksRequest(@JsonProperty("tasks") List<CreateTaskRequest> tasks) {
+        this.tasks = tasks;
```
Comment (cmccabe) [inline tools/src/main/java/org/apache/kafka/trogdor/rest/CreateMultipleTasksRequest.java]: We should handle the case where this is null (by treating it as empty)

File: tools/src/test/java/org/apache/kafka/trogdor/coordinator/CoordinatorTest.java [hunk 5]
```diff
@@ -102,9 +181,13 @@ public void testCreateTask() throws Exception {
                     build()).
                 waitFor(cluster.coordinatorClient());
 
-            // Re-creating a task with the same arguments is not an error.
-            cluster.coordinatorClient().createTask(
-                new CreateTaskRequest("foo", fooSpec));
+            // Re-creating a task with the same arguments gives a RequestConflictException.
+            try {
+                cluster.coordinatorClient().createTask(new CreateTaskRequest("foo", fooSpec));
+                fail("Expected to get an exception when re-creating the same task");
```
Comment (cmccabe) [inline tools/src/test/java/org/apache/kafka/trogdor/coordinator/CoordinatorTest.java:235]: Re-creating the same task should not cause an exception.  Requests should be idempotent (so that we can just retransmit requests that got lost.)

File: tools/src/main/java/org/apache/kafka/trogdor/coordinator/TaskManager.java [hunk 6]
```diff
@@ -294,25 +298,82 @@ private JsonNode getCombinedStatus() {
...
+    private ManagedTask createTask(TaskDetail task) throws Throwable {
         try {
-            executor.submit(new CreateTask(id, spec)).get();
+            if (task.id.isEmpty()) {
+                throw new InvalidRequestException("Invalid empty ID in createTask request.");
+            }
+            ManagedTask duplicateTask = tasks.get(task.id);
+            if (duplicateTask != null) {
+                log.info("Task {} already exists with spec {}", duplicateTask.id, duplicateTask.spec);
+                throw new RequestConflictException("Task ID " + duplicateTask.id + " already exists");
```
Comment (cmccabe) [inline tools/src/main/java/org/apache/kafka/trogdor/coordinator/TaskManager.java]: I commented earlier about this, but re-creating the same task with the same spec should not cause an exception.  (It should just do nothing.)

Think about a scenario where the server receives a request, acts on it, but then the response gets lost.  When the client re-sends the same request, the response should be "success".  This allows us to be robust in the face of transient network problems.

File: tools/src/main/java/org/apache/kafka/trogdor/coordinator/TaskManager.java (hunk 2 above)
Comment (stanislavkozlovski) [reply tools/src/main/java/org/apache/kafka/trogdor/coordinator/TaskManager.java]: That is handled in `TaskManager#createTask()`. In the worst case, we'd create N-1 tasks and throw an error on the last one, but none would get scheduled. Do you think it's worth to fail early?

File: tools/src/main/java/org/apache/kafka/trogdor/coordinator/TaskManager.java (hunk 3 above)
Comment (stanislavkozlovski) [reply tools/src/main/java/org/apache/kafka/trogdor/coordinator/TaskManager.java]: Oh, nice catch. I think we should re-raise so it gets propagated and returned as a 400 to the caller. It's also not very atomic right now, is it?

File: tools/src/test/java/org/apache/kafka/trogdor/coordinator/CoordinatorTest.java (hunk 5 above)
Comment (stanislavkozlovski) [reply tools/src/test/java/org/apache/kafka/trogdor/coordinator/CoordinatorTest.java:235]: I'm not sure. This is still idempotent in the sense that a task isn't created. I thought it's somewhat confusing to return a `201 CREATED` when none was created. If you think it's alright, I'll change it

File: tools/src/main/java/org/apache/kafka/trogdor/coordinator/TaskManager.java (hunk 6 above)
Comment (stanislavkozlovski) [reply tools/src/main/java/org/apache/kafka/trogdor/coordinator/TaskManager.java]: We are robust either way, the task is created. The question is how we want to inform the caller of what has happened - a `409` or `201` both do the same thing, they tell him that a task with that ID already exists on the server

File: tools/src/main/java/org/apache/kafka/trogdor/rest/CreateMultipleTasksRequest.java (hunk 4 above)
Comment (stanislavkozlovski) [reply tools/src/main/java/org/apache/kafka/trogdor/rest/CreateMultipleTasksRequest.java]: I think that we should just make the `tasks` field required. When the user doesn't provide it he'll get a 400 - it doesn't make sense to accept an empty json.
He can still send an empty list and that would return 201

File: tools/src/main/java/org/apache/kafka/trogdor/coordinator/CoordinatorClient.java [hunk 7]
```diff
@@ -125,6 +126,13 @@ public void createTask(CreateTaskRequest request) throws Exception {
         resp.body();
     }
 
+    public void createMultipleTasks(CreateMultipleTasksRequest request) throws Exception {
+        HttpResponse<Empty> resp =
+            JsonRestServer.httpRequest(log, url("/coordinator/task/mass_create"), "POST",
```
Comment (cmccabe) [inline tools/src/main/java/org/apache/kafka/trogdor/coordinator/CoordinatorClient.java]: `/coordinator/task/creates` seems better than "mass_create".

File: tools/src/main/java/org/apache/kafka/trogdor/coordinator/CoordinatorRestResource.java [hunk 8]
```diff
@@ -76,9 +78,16 @@ public CoordinatorStatusResponse status() throws Throwable {
...
+    }
+
+    @POST
+    @Path("/task/mass_create")
```
Comment (cmccabe) [inline tools/src/main/java/org/apache/kafka/trogdor/coordinator/CoordinatorRestResource.java]: `/task/creates`

File: tools/src/main/java/org/apache/kafka/trogdor/coordinator/CoordinatorClient.java [hunk 9]
```diff
@@ -259,6 +267,11 @@ public static void main(String[] args) throws Exception {
                 readValue(res.getString("create_task"), CreateTaskRequest.class);
             client.createTask(req);
             System.out.printf("Sent CreateTaskRequest for task %s.", req.id());
+        } else if (res.getString("create_multiple_task") != null) {
```
Comment (cmccabe) [inline tools/src/main/java/org/apache/kafka/trogdor/coordinator/CoordinatorClient.java]: Perhaps `create_tasks` rather than `create_multiple_tasks`?  Also, did you add a command line option for this?

File: tools/src/main/java/org/apache/kafka/trogdor/coordinator/TaskManager.java (hunk 6 above)
Comment (cmccabe) [reply tools/src/main/java/org/apache/kafka/trogdor/coordinator/TaskManager.java]: Check TROGDOR.md.

> All Trogdor RPCs are idempotent except the shutdown requests.  Sending an idempotent RPC twice in a row has the same effect as sending the RPC once.

Because the request is idempotent, sending it twice has the same effect, including the same result code.

File: tools/src/main/java/org/apache/kafka/trogdor/coordinator/TaskManager.java (hunk 3 above)
Comment (cmccabe) [reply tools/src/main/java/org/apache/kafka/trogdor/coordinator/TaskManager.java]: I don't think it needs to be atomic.  What I would suggest is something like:

1. Do basic validation that no task IDs are null, the JSON is well-formed, etc.  Most of this is done by jetty automatically, but we do need to manually check that the IDs aren't null, and that there is at least one task in the array.
2. Atomically, check to make sure that every task ID that we're trying to create either:
a. doesn't already exist, or
b. does exist, and with an identical task spec to what we sent
If anything doesn't match, we throw an exception and return an HTTP error for everything.
3. Create each task.  If there is an error during task creation, the task is just created in DONE state with error set to the error.
4. Return 201 CREATED.

Step 3 doesn't have to be atomic -- some parts of the request can succeed and some can fail.  This is fine, because the user is supposed to check the error state of the tasks periodically anyway.

If you want to get fancy with validation we could validate that all the provided class names actually exist and can be instantiated during step 2, and return an HTTP error if not.  But that is probably easier to do in a follow-on change.

Comment (stanislavkozlovski) [reply tools/src/main/java/org/apache/kafka/trogdor/coordinator/TaskManager.java]: Okay agreed, that makes sense, especially for step 3. If a user sends 4 tasks and two of them are configured funnily, he should be able to see what's wrong with them properly

File: tools/src/main/java/org/apache/kafka/trogdor/rest/CreateMultipleTasksRequest.java (hunk 4 above)
Comment (cmccabe) [reply tools/src/main/java/org/apache/kafka/trogdor/rest/CreateMultipleTasksRequest.java]: I think it's easier just to treat null as empty.  Otherwise we have to worry about testing another code path that really doesn't add any value to the user

File: tools/src/test/java/org/apache/kafka/trogdor/coordinator/CoordinatorTest.java (hunk 5 above)
Comment (cmccabe) [reply tools/src/test/java/org/apache/kafka/trogdor/coordinator/CoordinatorTest.java:235]: Idempotent means the same result is returned for the same input.  Not just a similar one.

**Comment by stanislavkozlovski** [comment]
Sorry about the checkstyle failure noise, my local `./gradlew check`/`./gradlew test` did not catch anything

**Comment by stanislavkozlovski** [comment]
Rebased with trunk

File: tools/src/test/java/org/apache/kafka/trogdor/coordinator/CoordinatorTest.java (hunk 5 above)
Comment (ijuma) [reply tools/src/test/java/org/apache/kafka/trogdor/coordinator/CoordinatorTest.java:235]: Technically, idempotence is used in multiple contexts and your definition is stricter than usual for HTTP. PUT is meant to be idempotent and the usual behaviour is to return 201 when an item is created and 200 or 204 when it is updated.

"If a new resource is created, the origin server MUST inform the user agent via the 201 (Created) response. If an existing resource is modified, either the 200 (OK) or 204 (No Content) response codes SHOULD be sent to indicate successful completion of the request."

https://www.w3.org/Protocols/rfc2616/rfc2616-sec9.html

File: tools/src/main/java/org/apache/kafka/trogdor/coordinator/TaskManager.java (hunk 6 above)
Comment (ijuma) [reply tools/src/main/java/org/apache/kafka/trogdor/coordinator/TaskManager.java]: See my other message, I don't think it has to have the same status code. However, I agree that it is unexpected to receive an error if the request succeeded. But a 200/204 are possible options.

File: tests/kafkatest/utils/util.py [hunk 10]
```diff
@@ -33,25 +33,29 @@ def _kafka_jar_versions(proc_string):
         - kafka-streams-1.0.0-SNAPSHOT.jar
         - kafka-streams-0.11.0.0-SNAPSHOT.jar
     """
+    def match(regex_pattern):
+        # return the result from the first regex group
+        return [result[0] if isinstance(result, tuple) else result
+                for result in re.findall(regex_pattern, proc_string)]
```
Comment (cmccabe) [inline tests/kafkatest/utils/util.py]: Is this change related to KAFKA-7638?

File: tools/src/main/java/org/apache/kafka/trogdor/coordinator/Coordinator.java [hunk 11]
```diff
@@ -91,7 +95,16 @@ public CoordinatorStatusResponse status() throws Exception {
     }
 
     public void createTask(CreateTaskRequest request) throws Throwable {
-        taskManager.createTask(request.id(), request.spec());
+        taskManager.createAndScheduleTask(new TaskManager.TaskDetail(request.id(), request.spec()));
+    }
+
+    public void createMultipleTasks(CreateMultipleTasksRequest request)
```
Comment (cmccabe) [inline tools/src/main/java/org/apache/kafka/trogdor/coordinator/Coordinator.java]: We can simplify this by just having a single function, `void createTasks(Map<String, TaskSpec> tasks)`

File: tools/src/main/java/org/apache/kafka/trogdor/rest/CreateMultipleTasksRequest.java [hunk 12]
```diff
@@ -0,0 +1,40 @@
+/*
```
Comment (cmccabe) [inline tools/src/main/java/org/apache/kafka/trogdor/rest/CreateMultipleTasksRequest.java]: `CreateTasksRequest` seems like a simpler name

File: tools/src/main/java/org/apache/kafka/trogdor/rest/CreateMultipleTasksRequest.java [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+import java.util.List;
+
+/**
+ * A request to the Trogdor coordinator to create mulitple tasks.
+ */
+public class CreateMultipleTasksRequest extends Message {
+    private final List<CreateTaskRequest> tasks;
```
Comment (cmccabe) [inline tools/src/main/java/org/apache/kafka/trogdor/rest/CreateMultipleTasksRequest.java]: I think this should just be a `Map<String, TaskSpec>`.  That's simpler

File: tools/src/test/java/org/apache/kafka/trogdor/coordinator/CoordinatorTest.java (hunk 5 above)
Comment (cmccabe) [reply tools/src/test/java/org/apache/kafka/trogdor/coordinator/CoordinatorTest.java:235]: I guess to me, the important thing here is that software that depends on Trogdor should not "see" a request that is retried as different from one which completed successfully the first time.  In order to do that, the return codes need to be the same, not just similar.  This is similar to how, in Kafka, we don't throw an exception or otherwise behave differently if we needed to send a ProduceRequest more than once.

File: tools/src/test/java/org/apache/kafka/trogdor/common/ExpectedTasks.java [hunk 14]
```diff
@@ -142,17 +144,23 @@ public ExpectedTasks addTask(ExpectedTask task) {
...
+            TasksResponse tasks;
+            try {
+                tasks = client.tasks(new TasksRequest(null, 0, 0, 0, 0, Optional.empty()));
+            } catch (Exception e) {
+                log.info("Unable to get coordinator tasks", e);
+                throw new RuntimeException(e);
+            }
+            StringBuilder errors = new StringBuilder();
+            if (tasks.tasks().size() < expected.size()) {
+                errors.append("We expected {} tasks but received {}", tasks.tasks().size(), expected.size());
```
Comment (cmccabe) [inline tools/src/test/java/org/apache/kafka/trogdor/common/ExpectedTasks.java]: I don't think this is necessary, and it makes the error log messages less useful.  Formerly we got specific details about what was missing; now, we just get a blanket error message that there aren't enough tasks.

File: tools/src/main/java/org/apache/kafka/trogdor/coordinator/TaskManager.java [hunk 15]
```diff
@@ -288,25 +291,103 @@ private JsonNode getCombinedStatus() {
...
 
     /**
-     * Create a task.
-     *
-     * @param id                    The ID of the task to create.
-     * @param spec                  The specification of the task to create.
+     * Creates and schedules a task.
      */
-    public void createTask(final String id, TaskSpec spec)
+    public void createAndScheduleTask(TaskDetail task)
```
Comment (cmccabe) [inline tools/src/main/java/org/apache/kafka/trogdor/coordinator/TaskManager.java]: Like I wrote above, we should just have one function that takes a map.  Then we don't need a special function for the case where there's only one task to create (can use Collections.singletonMap or similar if needed).

File: tools/src/main/java/org/apache/kafka/trogdor/coordinator/TaskManager.java (hunk 2 above)
Comment (cmccabe) [reply tools/src/main/java/org/apache/kafka/trogdor/coordinator/TaskManager.java]: Like I wrote earlier, this should just be a map, so duplicates should not be a problem.  I think it would be good to do all the validation here.  There's no reason not to do it and it makes things more robust if the code is re-arranged in the future.

File: tools/src/main/java/org/apache/kafka/trogdor/coordinator/TaskManager.java [hunk 16]
```diff
@@ -288,25 +291,103 @@ private JsonNode getCombinedStatus() {
...
+        if (tasks.stream().map(task -> task.id).collect(Collectors.toSet()).size() < tasks.size())
+            throw new RequestConflictException("Duplicate task IDs given");
+
+        for (TaskDetail task: tasks) {
+            try {
+                ManagedTask createdTask = createTask(task);
+                if (createdTask != null)
+                    managedTasks.add(createdTask);
+            } catch (RequestConflictException | InvalidRequestException e) {
+                throw e;
```
Comment (cmccabe) [inline tools/src/main/java/org/apache/kafka/trogdor/coordinator/TaskManager.java]: I thought I left a comment about this earlier.  Anyway, we don't want to leave things in a partial state if one thing fails.  If something fails, the task is in DONE state with an error, and every other task goes on as usual.

File: tools/src/main/java/org/apache/kafka/trogdor/rest/CreateMultipleTasksRequest.java (hunk 4 above)
Comment (cmccabe) [reply tools/src/main/java/org/apache/kafka/trogdor/rest/CreateMultipleTasksRequest.java]: So, following up on this discussion: the latest change is to add `required = true`.  However, this doesn't prevent someone from setting the value to to `null`.  See https://stackoverflow.com/questions/18320731/jackson-jsonpropertyrequired-true-doesnt-throw-an-exception  So we should still check for null and use an empty map here if needed.

Also, please make this unmodifiable using `Collections.unmodifiableMap`

File: tools/src/main/java/org/apache/kafka/trogdor/coordinator/CoordinatorRestResource.java [hunk 17]
```diff
@@ -79,9 +81,21 @@ public CoordinatorStatusResponse status() throws Throwable {
...
-    public Empty createTask(CreateTaskRequest request) throws Throwable {
+    public Response createTask(CreateTaskRequest request) throws Throwable {
         coordinator().createTask(request);
-        return Empty.INSTANCE;
+        return Response.status(201).entity(Empty.INSTANCE).build();
+    }
+
+    @POST
+    @Path("/task/creates")
+    public Response createMultipleTasks(CreateMultipleTasksRequest request) throws Throwable {
```
Comment (cmccabe) [inline tools/src/main/java/org/apache/kafka/trogdor/coordinator/CoordinatorRestResource.java]: should be `createTasks`

File: tools/src/main/java/org/apache/kafka/trogdor/coordinator/CoordinatorClient.java [hunk 18]
```diff
@@ -127,6 +128,13 @@ public void createTask(CreateTaskRequest request) throws Exception {
         resp.body();
     }
 
+    public void createMultipleTasks(CreateMultipleTasksRequest request) throws Exception {
```
Comment (cmccabe) [inline tools/src/main/java/org/apache/kafka/trogdor/coordinator/CoordinatorClient.java]: Should be `createTasks`

File: tools/src/main/java/org/apache/kafka/trogdor/coordinator/TaskManager.java [hunk 19]
```diff
@@ -288,25 +291,83 @@ private JsonNode getCombinedStatus() {
...
+                throw new InvalidRequestException("Invalid empty ID in createTask request.");
+
+            TaskSpec spec = idAndSpec.getValue();
+            ManagedTask duplicateTask = this.tasks.get(id);
+            if (duplicateTask != null) {
+                if (!duplicateTask.spec.equals(spec))
+                    throw new RequestConflictException("Task ID " + id + " already " +
+                        "exists, and has a different spec " + duplicateTask.spec);
+                log.info("Task {} already exists with spec {}", id, spec);
+                if (tasks.size() == 1)
```
Comment (cmccabe) [inline tools/src/main/java/org/apache/kafka/trogdor/coordinator/TaskManager.java]: Cleaner to just check if `tasks.isEmpty` after the loop is over.

File: tools/src/main/java/org/apache/kafka/trogdor/coordinator/TaskManager.java [hunk 20]
```diff
@@ -316,42 +377,13 @@ public void createTask(final String id, TaskSpec spec)
...
-            String failure = null;
+        public ManagedTask call() throws Exception {
             try {
-                controller = spec.newController(id);
-            } catch (Throwable t) {
-                failure = "Failed to create TaskController: " + t.getMessage();
+                TaskController controller = spec.newController(id);
+                return new ManagedTask(id, spec, controller, TaskStateType.PENDING);
+            } catch (Exception t) {
+                throw new Exception("Failed to create TaskController: " + t.getMessage());
```
Comment (cmccabe) [inline tools/src/main/java/org/apache/kafka/trogdor/coordinator/TaskManager.java]: This shouldn't throw an exception.  The existing logic where we fail the task is what we want to do here.

File: tools/src/main/java/org/apache/kafka/trogdor/coordinator/TaskManager.java [hunk 21]
```diff
@@ -288,25 +291,83 @@ private JsonNode getCombinedStatus() {
...
+        long delayMs = task.startDelayMs(time.milliseconds());
+        task.startFuture = scheduler.schedule(executor, new RunTask(task), delayMs);
+        return delayMs;
+    }
+
+    /**
+     * @return a #{@link ManagedTask} instance or #{@code null} if there was an error creating the task
+     *  or if the task already exists
+     */
+    private ManagedTask createTask(String id, TaskSpec spec) throws Throwable {
```
Comment (cmccabe) [inline tools/src/main/java/org/apache/kafka/trogdor/coordinator/TaskManager.java]: I don't think we really need this function any more... we can just submit to the executor from the other function.

File: tools/src/main/java/org/apache/kafka/trogdor/coordinator/TaskManager.java [hunk 22]
```diff
@@ -288,25 +291,83 @@ private JsonNode getCombinedStatus() {
...
+                id, spec, e.getMessage());
+            return null;
         }
     }
 
     /**
      * Handles a request to create a new task.  Processed by the state change thread.
      */
-    class CreateTask implements Callable<Void> {
+    class CreateTask implements Callable<ManagedTask> {
```
Comment (cmccabe) [inline tools/src/main/java/org/apache/kafka/trogdor/coordinator/TaskManager.java]: I don't think this `Callable` should return `ManagedTask`.  `ManagedTask` is an internal data structure that should only be accessed from the state change thread.

File: tools/src/main/java/org/apache/kafka/trogdor/coordinator/Coordinator.java [hunk 23]
```diff
@@ -91,7 +95,12 @@ public CoordinatorStatusResponse status() throws Exception {
     }
 
     public void createTask(CreateTaskRequest request) throws Throwable {
-        taskManager.createTask(request.id(), request.spec());
+        taskManager.createAndScheduleTasks(Collections.singletonMap(request.id(), request.spec()));
+    }
+
+    public void createMultipleTasks(CreateTasksRequest request)
```
Comment (cmccabe) [inline tools/src/main/java/org/apache/kafka/trogdor/coordinator/Coordinator.java]: Should be `createTasks`

File: tools/src/main/java/org/apache/kafka/trogdor/coordinator/CoordinatorRestResource.java [hunk 24]
```diff
@@ -79,9 +81,19 @@ public CoordinatorStatusResponse status() throws Throwable {
...
-    public Empty createTask(CreateTaskRequest request) throws Throwable {
+    public Response createTask(CreateTaskRequest request) throws Throwable {
         coordinator().createTask(request);
-        return Empty.INSTANCE;
+        return Response.status(201).entity(Empty.INSTANCE).build();
+    }
+
+    @POST
+    @Path("/task/creates")
+    public Response createMultipleTasks(CreateTasksRequest request) throws Throwable {
```
Comment (cmccabe) [inline tools/src/main/java/org/apache/kafka/trogdor/coordinator/CoordinatorRestResource.java]: should be `createTasks`

File: tools/src/main/java/org/apache/kafka/trogdor/coordinator/TaskManager.java [hunk 25]
```diff
@@ -288,19 +289,44 @@ private JsonNode getCombinedStatus() {
...
+        Map<String, TaskSpec> tasks = new HashMap<>(givenTasks);
+        Iterator<Map.Entry<String, TaskSpec>> it = tasks.entrySet().iterator();
+        while (it.hasNext()) {
+            Map.Entry<String, TaskSpec> idAndSpec = it.next();
+            String id = idAndSpec.getKey();
+            if (id.isEmpty())
+                throw new InvalidRequestException("Invalid empty ID in createTask request.");
+
+            TaskSpec spec = idAndSpec.getValue();
+            ManagedTask duplicateTask = this.tasks.get(id);
```
Comment (cmccabe) [inline tools/src/main/java/org/apache/kafka/trogdor/coordinator/TaskManager.java]: `TaskManager#tasks` has to be accessed through the state change thread.  It can't be accessed here by incoming requests since there is no lock or synchronization.  Probably the easiest thing to do is to have a CreateTasks runnable which does what you want.

File: tools/src/main/java/org/apache/kafka/trogdor/coordinator/TaskManager.java (hunk 16 above)
Comment (cmccabe) [reply tools/src/main/java/org/apache/kafka/trogdor/coordinator/TaskManager.java]: Hmm.  This still has the problem where things can be partially applied, because we have a bunch of `CreateTask` runnables being processed separately.  It would be easier to just have a single `CreateTasks` runnable and pass it the map.  Then the whole thing could fail with a `RequestConflictException` if any task had a conflict.
